rust/tools/authorization-logic/README.md

```diff
@@ -0,0 +1,209 @@
+says somePred(x)`.
+
+When writing facts or rules, the arguments to predicates can either be 
+constants or variables. Constants are written in quotes whereas variables are 
+not. For example,
+```
+"Alice" says somePred("x") :- condition("x").
+```
+gives a rule for proving somePred for just a particular argument, `"x"`, 
+whereas the same rule replaced with `x` can be applied to any argument about 
+which we can prove `"Alice" says condition(x)`.
+
+Principal names can also be used as arguments to predicates, and quotes are used
+indicate whether a principal is a constant or variable in the same way.
+
+Multiple rules/facts stated by the same principal can be grouped together using 
+curly braces, as in:
+```
+"Alice" says {
+    rule1(x) :- condition(x).
+    rule2(y) :- condition1(y), condition(x).
+    someFact(x).
+}
+```
+
+#### Delegation
+
+Delegation works by prepending a predicate on the left of a fact or rule with 
+`<principal_name> canSay`. Delegations can only appear on the left side of a 
+rule. As an example, the rule
+```
+"Alice" says "Bob" canSay pred(x) :- condition(x).
+```
+can be combined with `"Alice" says condition("y")` and `"Bob" says pred("y")` 
+to prove `Alice says pred("y")`.
+
+#### Attributes and canActAs
+
+Predicates can be used to represent attributes of principals or other 
+constants/variables using the syntax `principal_name predicate_name(<args>)`, 
+such as, `"Manager" hasTraining("interviewing")`. Attributes can appear in 
+facts or in the left or right of a rule. 
+
+Attributes can be passed from one principal to another using `canActAs`. If `A 
+canActAs B`, then `A` inherits all the attributes of `B`. For example,
+```
+"HR" says "Alice" canActAs "Manager".
+```
+can be combined with `"HR" says "Manager" hasTraining("interviewing")` to prove
+`"HR" says "Alice" hasTraining("interviewing")`. 
+
+Combining attributes with `canActAs` gives a way to define _roles_ which group 
+privileges together.
+
+#### Queries
+Queries test if it is possible to prove a specific predicate given the facts 
+and rules. Queries can be written using the syntax:
+```
+<query_name> = query <principal_name> says <predicate_name>(<args>)?
+```
+
+If the program contained queries (described below), then the there will also be
+a file called `<query_name>.csv` for each query. The query is true if the CSV
+contains one entry and is false otherwise.
+
+#### Importing and Exporting statements
+
+Statements can be imported/exported from potentially untrusted sources such as 
+remote machines or storage. To authenticate statements, they are digitally 
+signed using ECDSA signatures when exporting, and the signatures are checked 
+when importing. 
+
+To support exporting, a principal must be bound to a private key. Private and 
+public keys must be serialized in JSON. See 
+[tink-rust](https://github.com/project-oak/tink-rust) for a way to generate 
+key-pairs. Private keys must be serialized in plaintext, so care must be taken.
```

Comment:
I think there are cases where serializing the key in plaintext is right even for production uses. No matter what, eventually a private (or symmetric) key lives in plaintext somewhere. Tink provides APIs for key management services that store the key on AWS or google cloud, where some private key is stored in plaintext on their machines. This is probably better for many use-cases, but an offline device would not be able to contact remote servers (and a similar case can be made where connection is available but sporadic). In these cases, the system designer should keep the keys in plaintext locally, but do something else in the system design to keep the private key out of reach by the adversary (but what this something else is, depends on the system).

In short, I think even "eventually" we want to support both.

What if after "... so care must be taken" we have:

> Anyone with a principal's private key can write (possibly forged) signatures
> as this principal, so the system in which this
> policy language is used must ensure these keys are only accessible by the right
> party. Future versions of this software may also support key management
> services such as [Google Key Management](https://cloud.google.com/security-key-management)
> or [AWS Key Management Service](https://aws.amazon.com/kms/).

Reply:
Though do also let me know if there is another better practice for offline use-cases. (I may just be unaware).

Reply:
The preferred way for offline scenarios is to use hardware-backed security devices (where the private key is generated and used in the device and can never leave it or be exported or copied) such as a smart card, TPM, etc. I guess it depends on the level of security required for a scenario and the potential threat model, but I think that any system that stores cryptographic material in clear text on any form of persistent storage should be considered insecure for anything except the most basic of threat models (which I assume is why the tink feature to enabled this is called "insecure").

Seeing that this is only a research prototype for now, I think the updated wording is ok. In future it might be useful to add features to this library (such as `insecure`, `google-kms`, `aws-kms`) so that the user can choose how keysets can be persisted. This would then require the user to actively choose an insecure configuration if it is needed rather than the insecure configuration being the default.